src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,102 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elements;
+    private int size;
+
+    private void ensureCapacity() {
+        if (size == elements.length) {
+            int newCapacity = (int)(elements.length * 1.5);
+            elements = Arrays.copyOf(elements, newCapacity);
+        }
+    }
+
+    private void ensureCapacity(int DEFAULT_CAPACITY) {
+        if (DEFAULT_CAPACITY > elements.length) {
+            int newCapacity = elements.length + elements.length / 2;
+            if (newCapacity < DEFAULT_CAPACITY) {
+                newCapacity = DEFAULT_CAPACITY;
+            }
+            elements = Arrays.copyOf(elements, newCapacity);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        ensureCapacity();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
```

Comment:
The index check logic in the add(T value, int index) method is repeated in other methods such as get(int index), set(T value, int index), and remove(int index). Move this check to a separate private method to avoid code duplication.